### PR TITLE
Gitolite special chars

### PR DESCRIPTION
**Pull request by beyang**
Fixes https://github.com/sourcegraph/issues-uber/issues/151. This PR is best reviewed commit by commit at first, but there are inline comments about the mock/DI patterns in the overall PR.

I tried adopting dependency injection in two of the packages, along with using gomock to generate mocks/stubs for unit tests. DI (the way it is implemented here) results in a little bit more verbosity and indirection complexity, but I think it's worth it for the stronger unit tests. I also tried to write the tests in a way such that the boilerplate could be codegen'd at some point in the future.

### CODE REVIEW

File: pkg/extsvc/gitolite/repos.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+// Client is a client for the Gitolite API.
+//
+// IMPORTANT: in order to authenticate to the Gitolite API, the client must be invoked from a
+// service in an environment that contains a Gitolite-authorized SSH key. As of writing, only
+// gitserver meets this criterion. (I.e., only invoke this from gitserver.)
+type Client struct {
+	Host string
+
+	command Command
```
Comment (beyang) [inline pkg/extsvc/gitolite/repos.go]: `command` is an injected dependency.

File: cmd/gitserver/server/list-gitolite.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	client IGitoliteClient
+	config IConfig
+}
+
+var defaultGitolite = Gitolite{client: GitoliteClient{}, config: Config{}}
+
+// listGitolite is effectively a wrapper around gitolite.Client.ListRepos.  This must currently be
+// invoked from gitserver, because only gitserver has the SSH key needed to authenticate to the
+// Gitolite API.
+func (g Gitolite) listGitolite(ctx context.Context, gitoliteHost string, w http.ResponseWriter) {
```
Comment (beyang) [inline cmd/gitserver/server/list-gitolite.go]: `listGitolite` was originally a function, but in order to unit test, the dependencies should be mocked. The cleanest way to do that with dependency injection seems to be a struct (`Gitolite`) with fields for the dependencies to be injected.

File: cmd/gitserver/server/list-gitolite.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+func (c GitoliteClient) ListRepos(ctx context.Context, host string) ([]*gitolite.Repo, error) {
+	return gitolite.NewClient(host).ListRepos(ctx)
+}
+
+type Gitolite struct {
+	client IGitoliteClient
+	config IConfig
+}
+
+var defaultGitolite = Gitolite{client: GitoliteClient{}, config: Config{}}
```
Comment (beyang) [inline cmd/gitserver/server/list-gitolite.go]: Perhaps at some point the injected dependencies will also be injected throughout the entire function call graph, but that would require refactoring a lot of functions (and make it so the `conf` package no longer contains globals). For now, the dependency injection pattern is isolated to this package and the interface to external globals is this global variable.

**Comment by tsenart** [comment]
I'll review this tomorrow morning. Feeling under the weather so calling it a day.

File: pkg/extsvc/gitolite/repos.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		}
+	}
+
+	return repos, nil
+}
+
+// Command is an interface for invoking a shell command.
+type Command interface {
+	Output(ctx context.Context, name string, arg ...string) ([]byte, error)
+}
```
Comment (tsenart) [inline pkg/extsvc/gitolite/repos.go]: Exposing such a general purpose interface makes testing more laborious that it needs to be. What is the client really depending on? The return of the byte stream that needs to be decoded. Exposing the client to the implementation details of how that happens isn't strictly necessary.

Consider this alternative where the client isn't even aware of any SSH details.

```go
// Command represents a gitolite info command runner that lists repositories.
type Command interface {
    Info(context.Context) (io.Reader, error)
}

// SSHCommand implements the Command interface by running an SSH
// info command against a real gitolite server.
type SSHCommand struct {
    Host string
}

// Run implements the Command interface.
func (c SSHCommand) Info(ctx context.Context) (io.Reader, error) {
    buf := new(bytes.Buffer)
    cmd := exec.CommandContext(ctx, "ssh", c.Host, "info")
    cmd.Stdout = buf
    return buf, cmd.Run()
}
```

**Comment by beyang** [comment]
I ended up abandoning use of `gomock` and just writing more straightforward unit tests. Still kept the DI pattern where it made sense.

**Comment by tsenart** [comment]
> I ended up abandoning use of gomock and just writing more straightforward unit tests. Still kept the DI pattern where it made sense.

Looks much better now!

**Pull request merged**